shared/src/ledger/native_vp/masp.rs

```diff
@@ -0,0 +1,296 @@
+                    continue;
+                }
+                if !valid_transfer_amount(
+                    out.value,
+                    denom.denominate(&transfer.amount.amount),
+                ) {
+                    return Ok(false);
+                }
+
+                let (_transp_asset, transp_amt) = convert_amount(
+                    epoch,
+                    &transfer.token,
+                    transfer.amount.amount,
+                    denom,
+                );
+
+                // Non-masp destinations subtract from transparent tx pool
+                transparent_tx_pool -= transp_amt;
+
+                // Satisfies 4.
+                let target_enc = transfer.target.serialize_to_vec();
+
+                let hash = ripemd::Ripemd160::digest(sha2::Sha256::digest(
+                    &target_enc,
+                ));
+
+                if <[u8; 20]>::from(hash) != out.address.0 {
+                    tracing::debug!(
+                        "the public key of the output account does not match \
+                         the transfer target"
+                    );
+                    return Ok(false);
+                }
+                valid_count += 1;
+            }
+            // one or more of the denoms in the batch failed to verify
+            // the asset derivation.
+            if valid_count != out_length {
+                return Ok(false);
+            }
+        } else {
+            // Handle shielded output
+            // The following boundary conditions must be satisfied
+            // 1. Zero transparent output
+
+            // Satisfies 1.
+            if let Some(transp_bundle) = shielded_tx.transparent_bundle() {
+                if !transp_bundle.vout.is_empty() {
+                    tracing::debug!(
+                        "Transparent output to a transaction from the masp \
+                         must be 0 but is {}",
+                        transp_bundle.vout.len()
+                    );
+                    return Ok(false);
+                }
+            }
+        }
+
+        match transparent_tx_pool.partial_cmp(&I128Sum::zero()) {
+            None | Some(Ordering::Less) => {
+                tracing::debug!(
+                    "Transparent transaction value pool must be nonnegative. \
+                     Violation may be caused by transaction being constructed \
+                     in previous epoch. Maybe try again."
+                );
+                // Section 3.4: The remaining value in the transparent
+                // transaction value pool MUST be nonnegative.
+                return Ok(false);
+            }
+            Some(Ordering::Greater) => {
+                tracing::debug!(
+                    "Transaction fees cannot be paid inside MASP transaction."
+                );
+                return Ok(false);
+            }
+            _ => {}
+        }
+        // Verify the proofs.
+        Ok(verify_shielded_tx(&shielded_tx))
```

Comment:
It would be nice to benchmark this function because I think this is the most computationally expensive one and charge some gas before calling it in here